File: dotnet/v2.6/Misc/AuthenticateUsingServiceAccount.cs
/*
 * Copyright 2015 Google Inc
 *
 * Licensed under the Apache License, Version 2.0(the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using Google.Apis.Auth.OAuth2;
using Google.Apis.Dfareporting.v2_6;
using Google.Apis.Dfareporting.v2_6.Data;
using Google.Apis.Json;
using Google.Apis.Services;
using System;
using System.IO;

namespace DfaReporting.Samples {
  /// <summary>
  /// This example shows how to authenticate and make a basic request using a service account.
  /// </summary>
  class AuthenticateUsingServiceAccount : SampleBase {
    /// <summary>
    /// Returns a description about the code example.
    /// </summary>
    public override string Description {
      get {
        return "This example shows how to authenticate and make a basic request" +
            " using a service account.\n";
      }
    }

    /// <summary>
    /// Main method, to run this code example as a standalone application.
    /// </summary>
    /// <param name="args">The command line arguments.</param>
    public static void Main(string[] args) {
      SampleBase codeExample = new AuthenticateUsingServiceAccount();
      Console.WriteLine(codeExample.Description);
      codeExample.Run(null);
    }

    /// <summary>
    /// Run the code example.
    /// </summary>
    /// <param name="service">Unused</param>
    public override void Run(DfareportingService service) {
      string emailToImpersonate = _T("ENTER_EMAIL_TO_IMPERSONATE_HERE");
      string pathToJsonFile = _T("ENTER_PATH_TO_JSON_FILE_HERE");

      // Create the Dfareporting service using service account credentials.
      service = new DfareportingService(new BaseClientService.Initializer {
        HttpClientInitializer =
            getServiceAccountCredential(pathToJsonFile, emailToImpersonate),
        ApplicationName = "DFA/DCM Reporting and Trafficking API Samples"
      });

      UserProfileList profiles = service.UserProfiles.List().Execute();

      foreach (UserProfile profile in profiles.Items) {
        Console.WriteLine("Found user profile with ID {0} and name \"{1}\".",
            profile.ProfileId, profile.UserName);
      }
    }

    private ServiceAccountCredential getServiceAccountCredential(String pathToJsonFile,
      String emailToImpersonate) {
      // Load and deserialize the specified JSON file.
      JsonCredentialParameters parameters;
      using (Stream json = new FileStream(pathToJsonFile, FileMode.Open, FileAccess.Read)) {
        parameters = NewtonsoftJsonSerializer.Instance.Deserialize<JsonCredentialParameters>(json);
      }

      // Generate a ServiceAccountCredential object with the correct scopes and impersonation info.
      return new ServiceAccountCredential(
         new ServiceAccountCredential.Initializer(parameters.ClientEmail) {
           Scopes = new[] {
             DfareportingService.Scope.Dfareporting,
             DfareportingService.Scope.Dfatrafficking,
             DfareportingService.Scope.Ddmconversions
           },
           User = emailToImpersonate
         }.FromPrivateKey(parameters.PrivateKey));
    }
  }
}

File: java/v2.6/src/main/java/com/google/api/services/samples/dfareporting/misc/AuthenticateUsingServiceAccount.java
// Copyright 2014 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.services.samples.dfareporting.misc;

import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
import com.google.api.services.dfareporting.Dfareporting;
import com.google.api.services.dfareporting.DfareportingScopes;
import com.google.api.services.dfareporting.model.UserProfileList;

import java.io.FileInputStream;

/**
 * This example shows how to authenticate and make a basic request using a service account.
 */
public class AuthenticateUsingServiceAccount {
  private static final String PATH_TO_JSON_FILE = "ENTER_PATH_TO_JSON_FILE_HERE";
  private static final String EMAIL_TO_IMPERSONATE = "ENTER_EMAIL_TO_IMPERSONATE_HERE";

  private static Credential getServiceAccountCredential(String pathToJsonFile,
      String emailToImpersonate) throws Exception {
    // Generate a credential object from the specified JSON file.
    GoogleCredential credential = GoogleCredential.fromStream(new FileInputStream(pathToJsonFile));

    // Update the credential object with appropriate scopes and impersonation info.
    credential = new GoogleCredential.Builder()
        .setTransport(credential.getTransport())
        .setJsonFactory(credential.getJsonFactory())
        .setServiceAccountId(credential.getServiceAccountId())
        .setServiceAccountPrivateKey(credential.getServiceAccountPrivateKey())
        .setServiceAccountScopes(DfareportingScopes.all())
        // Set the email of the user you are impersonating (this can be yourself).
        .setServiceAccountUser(emailToImpersonate)
        .build();

    return credential;
  }

  public static void runExample(Dfareporting reporting) throws Exception {
    // Retrieve and print all user profiles for the current authorized user.
    UserProfileList profiles = reporting.userProfiles().list().execute();
    for (int i = 0; i < profiles.getItems().size(); i++) {
      System.out.printf("%d) %s%n", i + 1, profiles.getItems().get(i).getUserName());
    }
  }

  public static void main(String[] args) throws Exception {
    // Build service account credential.
    Credential credential = getServiceAccountCredential(PATH_TO_JSON_FILE, EMAIL_TO_IMPERSONATE);

    Dfareporting reporting = new Dfareporting.Builder(credential.getTransport(),
        credential.getJsonFactory(), credential).setApplicationName("dfareporting-java-samples")
        .build();

    runExample(reporting);
  }
}
